R/bambu-processReads_utilityConstructReadClasses.R

```diff
@@ -356,23 +356,63 @@ initiateHitsDF <- function(hitsWithin, grangesReference, stranded) {
-    GENEID[newGeneSet] <- newGeneIds
-    return(data.frame(GENEID=GENEID, novelGene = newGeneSet))
+#' @param min.exonOverlap minimum length of overlap to be assigned to a gene
+#' @param fusionMode if TRUE will assign multiple GENEIDs to fusion transcripts, separated by ":"
+#' @param maxChainIteration the number of intermediate novel transcripts which will be used to assign gene ids, default: 3
+assignGeneIds <-  function(grl, annotations, min.exonOverlap = 10, fusionMode = FALSE, maxChainIteration = 3) {
+  strandedRanges <- as.logical(getStrandFromGrList(grl)!='*')
+  mcols(grl)$GENEID =rep(NA, length(grl))
+  if(length(annotations)>0) {
+    if(any(grepl('\\:', mcols(annotations)$GENEID)) & fusionMode) {
+      warning('GENEID contains ":" symbol, fusion transcript assignments might be incorrect')
+    }
+    mcols(grl)$GENEID[strandedRanges] <- assignGeneIdsByReference(grl[strandedRanges], annotations,
+                                          min.exonOverlap = min.exonOverlap,
+                                          fusionMode = fusionMode) 
+    #iteratively assign gene ids for stranded granges
+    newGeneSet <- is.na(mcols(grl)$GENEID) & strandedRanges
+    referenceGeneSet <- !is.na(mcols(grl)$GENEID) & strandedRanges
+    mcols(grl)$GENEID[newGeneSet] <- assignGeneIdsByReference(grl[newGeneSet], grl[referenceGeneSet],
+                                                                  min.exonOverlap = min.exonOverlap,
+                                                                  fusionMode = FALSE)  # fusion assignment is only done based on original annotations
+    chainCount <- 1  # first iteration is outside of while loop
+    while(any(!is.na(mcols(grl)$GENEID[newGeneSet])) & chainCount < maxChainIteration) {
+      referenceGeneSet <- newGeneSet & !is.na(mcols(grl)$GENEID) & strandedRanges
+      newGeneSet <- is.na(mcols(grl)$GENEID) & strandedRanges
+      mcols(grl)$GENEID[newGeneSet] <- assignGeneIdsByReference(grl[newGeneSet], grl[referenceGeneSet],
+                                                                min.exonOverlap = min.exonOverlap,
+                                                                fusionMode = FALSE) 
+      chainCount <- chainCount +1
+      }
+  }
+  newGeneSet <- is.na(mcols(grl)$GENEID) & strandedRanges
+  newGeneIds <- assignGeneIdsNoReference(grl[newGeneSet])
+  mcols(grl)$GENEID[newGeneSet] <- newGeneIds
+  if(any(!strandedRanges)) {
+    mcols(grl)$GENEID[!strandedRanges] <- assignGeneIdsByReference(grl[!strandedRanges], 
+                                                                   grl[!is.na(mcols(grl)$GENEID)],
+                                                       min.exonOverlap = min.exonOverlap,
+                                                       fusionMode = FALSE) 
+    if(any(is.na(mcols(grl)$GENEID))) {
+      mcols(grl)$GENEID[is.na(mcols(grl)$GENEID)] <- assignGeneIdsNoReference(grl[is.na(mcols(grl)$GENEID)])
```

Comment:
This assigns already preexisting gene ids from the first call to unrelated read classes

Reply:
solved in [db8ddcf](https://github.com/GoekeLab/bambu/pull/260/commits/db8ddcfc2e260f36945733e56b7ee0778a44e202):
- unstranded genes get now gene Ids "gene.unstranded.id"
related: to rely less on gene names to define novel gene ids, bambu now uses the novelGene column in the rowData object [8ec7abb](https://github.com/GoekeLab/bambu/pull/260/commits/8ec7abbf8c48c5353dbd03564f6b6890b7142601)